resolwe/observers/models.py

```diff
@@ -0,0 +1,144 @@
+    # Primary key of the observed resource (null if watching the whole table)
+    object_id = models.PositiveIntegerField(null=True)
+
+    change_type = models.CharField(choices=CHANGE_TYPES, max_length=6)
+
+    class Meta:
+        """Meta parameters for the Observer model."""
+
+        unique_together = ("content_type", "object_id", "change_type")
+        indexes = [
+            models.Index(name="idx_observer_object_id", fields=["object_id"]),
+        ]
+
+    @classmethod
+    def get_interested(cls, content_type, object_id=None, change_type=None):
+        """Find all observers watching for changes of a given item/table."""
+        query = Q(content_type=content_type)
+        if change_type is not None:
+            query &= Q(change_type=change_type)
+        query &= Q(object_id=object_id) | Q(object_id__isnull=True)
+        return cls.objects.filter(query)
+
+    @classmethod
+    def observe_instance_changes(cls, instance, change_type):
+        """Handle a notification about an instance change."""
+
+        observers = Observer.get_interested(
+            change_type=change_type,
+            content_type=ContentType.objects.get_for_model(type(instance)),
+            object_id=instance.pk,
+        )
+
+        # Forward the message to the appropriate groups.
+        for subscriber in Subscription.objects.filter(observers__in=observers):
+            has_permission = (
+                type(instance)
+                .objects.filter(pk=instance.pk)
+                .filter_for_user(user=subscriber.user)
+                .exists()
+            )
+            if not has_permission:
+                continue
+
+            # Register on_commit callbacks to send the signals.
+            Subscription.notify(subscriber.session_id, instance, change_type)
+
+    def __str__(self):
+        """Format the object representation."""
+        return f"content_type={self.content_type} object_id={self.object_id} change={self.change_type}"
+
+
+class Subscription(models.Model):
+    """Subscription to several observers."""
+
+    observers = models.ManyToManyField("Observer", related_name="subscriptions")
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+    )
+    created = models.DateTimeField(auto_now_add=True)
+
+    # ID of the websocket session (can have multiple observers)
+    session_id = models.CharField(max_length=100)
+    # Unique ID for the client to remember which subscription a signal belongs to
+    subscription_id = models.UUIDField(
+        unique=True, default=get_random_uuid, editable=False
+    )
+
+    class Meta:
+        """Meta parameters for the Subscription model."""
+
+        indexes = [
+            models.Index(name="idx_subscription_session_id", fields=["session_id"]),
+        ]
+
+    def subscribe(self, content_type, resource_ids, change_types):
+        """Assign self to multiple observers at once."""
+        for id in resource_ids:
```

Comment:
I think it would be best to stick to the same terminology: if you are using `object_id` in the observer, use it also here (instead of using `resource_id`).